### PR TITLE
Lending iterator

### PR DESCRIPTION
**Pull request by dmitrii-ubskii**

### CODE REVIEW

File: concept/thing/thing_manager.rs [hunk 1]
```diff
@@ -102,15 +118,46 @@ impl<Snapshot: ReadableSnapshot> ThingManager<Snapshot> {
...
+        player: &impl ObjectAPI<'o>,
+    ) -> impl for<'x> LendingIterator<Item<'x> = Result<Relation<'x>, ConceptReadError>> {
+        let prefix = ThingEdgeRolePlayer::prefix_reverse_from_player(player.vertex());
+        let snapshot_iterator =
+            snapshot.iterate_range(KeyRange::new_within(prefix, Prefix::EdgeRolePlayer.fixed_width_keys()));
+        RelationRoleIterator::new(snapshot_iterator).map::<Result<Relation<'_>, _>, _>(|res| {
+            let (rel, _, _) = res?;
+            Ok(rel)
+        })
+    }
```
Comment (dmitrii-ubskii) [inline concept/thing/thing_manager.rs:141]: Showcase

File: storage/snapshot/iterator.rs
Comment (dmitrii-ubskii) [inline storage/snapshot/iterator.rs:1]: Another highlight

File: common/lending_iterator/lending_iterator.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+impl<K: ?Sized, LI> Seekable<K> for Peekable<LI>
+where
+    LI: Seekable<K>,
+{
+    fn seek(&mut self, key: &K) {
+        if self.item.is_some() {
+            todo!()
+        }
+        self.item = None;
+        self.iter.seek(key)
```
Comment (flyingsilverfin) [inline common/lending_iterator/lending_iterator.rs:152]: Let's establish the convention for all `seek` operations: Seeking into the previous iterator states is an illegal states. This is sort of a strict behaviour, but aligns with how we will be using seek() throughout TypeDB.

File: concept/thing/attribute.rs [hunk 3]
```diff
@@ -255,7 +258,7 @@ impl<'a, Snapshot: ReadableSnapshot, const A_PS: usize, const H_PS: usize> Attri
         }
     }
 
-    fn iter_next(&mut self) -> Option<Result<(StorageKeyReference<'_>, ByteReference<'_>), ConceptReadError>> {
+    fn iter_next(&mut self) -> Option<Result<(StorageKey<'_, 40>, Bytes<'_, 64>), ConceptReadError>> {
```
Comment (flyingsilverfin) [inline concept/thing/attribute.rs:261]: Let's use the two constants throughout

File: storage/keyspace/iterator.rs [hunk 4]
```diff
@@ -4,160 +4,176 @@
...
-        // read_opts.set_prefix_same_as_start(true);
-        let read_opts = keyspace.new_read_options();
-        let raw_iterator: DBRawIteratorWithThreadMode<'a, DB> = keyspace.kv_storage.raw_iterator_opt(read_opts);
+    use lending_iterator::{LendingIterator, Seekable};
+    use speedb::DBRawIterator;
 
-        KeyspaceRangeIterator { keyspace_name: keyspace.name(), range, iterator: raw_iterator, state: State::Init }
+    pub(super) struct DBIterator {
+        iterator: DBRawIterator<'static>,
+        item: Option<Result<(&'static [u8], &'static [u8]), speedb::Error>>,
```
Comment (flyingsilverfin) [inline storage/keyspace/iterator.rs:26]: Leave a comment saying this isn't actually static

File: storage/keyspace/iterator.rs [hunk 5]
```diff
@@ -4,160 +4,176 @@
...
+    }
+}
+
+impl Seekable<[u8]> for KeyspaceRangeIterator {
+    fn seek(&mut self, key: &[u8]) {
+        self.iterator.seek(key);
+    }
+}
+
+impl KeyspaceRangeIterator {
```
Comment (flyingsilverfin) [inline storage/keyspace/iterator.rs:166]: Delete

**Review by flyingsilverfin — APPROVED** [review]

**Pull request merged**